Tests/BSGStorageMigratorTests.m

```diff
@@ -0,0 +1,196 @@
+        [c stringByAppendingPathComponent:@"bugsnag/state/system_state.json"]: [r stringByAppendingPathComponent:@"system_state.json"],
+    };
+}
+
+- (void)setUp {
+    NSFileManager *fm = [NSFileManager defaultManager];
+    NSBundle *bundle = [NSBundle bundleForClass:[self class]];
+    NSString *cachesPath = [self getCachesDir];
+    NSError *err = nil;
+
+    // Test case app data is kept in a globally shared location, so make sure it's cleared.
+    [fm removeItemAtPath:[self getV1RootDir] error:nil];
+    for(NSString* path in [self getFiles]) {
+        [fm removeItemAtPath:path error:nil];
+    }
+    [fm removeItemAtPath:[cachesPath stringByAppendingPathComponent:@"bugsnag"] error:nil];
+    [fm removeItemAtPath:[cachesPath stringByAppendingPathComponent:@"Sessions"] error:nil];
+    [fm removeItemAtPath:[cachesPath stringByAppendingPathComponent:@"KSCrashReports"] error:nil];
+    [fm removeItemAtPath:[cachesPath stringByAppendingPathComponent:@"bsg_kvstore"] error:nil];
+
+    // Now copy the faked app data across from the test fixture data.
+    NSString *srcPath = [[bundle resourcePath] stringByAppendingPathComponent:@"v0_files/Caches"];
+    for(NSString *file in [fm contentsOfDirectoryAtPath:srcPath error:&err]) {
+        if (![fm copyItemAtPath:[srcPath stringByAppendingPathComponent:file]
+                         toPath:[cachesPath stringByAppendingPathComponent:file]
+                          error:&err]) {
+            XCTFail(@"%@", err);
+        }
+    }
+    if(err != nil) {
+        XCTFail(@"%@", err);
+    }
+}
+
+- (void)testMigrateFiles {
+    NSDictionary *dirs = [self getDirs];
+    NSDictionary *files = [self getFiles];
+
+    NSFileManager *fm = [NSFileManager defaultManager];
+    BOOL isDir = false;
+
+    // Make sure pre-state is as we expect.
+    for(NSString *path in files) {
+        if(![fm fileExistsAtPath:path isDirectory:&isDir]) {
+            XCTFail(@"Expected path to exist: %@", path);
+        }
+        if(isDir) {
+            XCTFail(@"Expected a file: %@", path);
+        }
+    }
+    for(NSString *path in dirs) {
+        if(![fm fileExistsAtPath:path isDirectory:&isDir]) {
+            XCTFail(@"Expected path to exist: %@", path);
+        }
+        if(!isDir) {
+            XCTFail(@"Expected a directory: %@", path);
+        }
+    }
+
+    // Migrate and check.
+    [BSGStorageMigratorV0V1 migrate];
+    [self postMigrateCheck];
+
+    // Make sure it's idempotent.
+    [BSGStorageMigratorV0V1 migrate];
+    [BSGStorageMigratorV0V1 migrate];
+    [BSGStorageMigratorV0V1 migrate];
+    [BSGStorageMigratorV0V1 migrate];
+    [BSGStorageMigratorV0V1 migrate];
+    [BSGStorageMigratorV0V1 migrate];
+    [self postMigrateCheck];
+
+    // Make sure the client still loads and files are still as we expect.
+    BugsnagConfiguration *configuration = [[BugsnagConfiguration alloc] initWithApiKey:DUMMY_APIKEY_32CHAR_1];
+    (void)[[BugsnagClient alloc] initWithConfiguration:configuration];
```

Comment:
I don't think it's a good idea to instantiate the BugsnagClient in the BSGStorageMigratorTests

BugsnagClient is quite a heavy object to instantiate, registers itself as an observer for various notifications, and writes to disk based on those notifications. Combined with the fact that BugsnagClient objects never deallocate (possible retain cycle) means that instantiating them in tests can cause weird issues.